tests/src/JIT/Performance/CodeQuality/Span/SpanBench.cs

```diff
@@ -447,82 +532,189 @@ public static void TestSpanDangerousGetPinnableReferenceByte(int length)
-            var span = new Span<string>(array);
-            foreach (var iteration in Benchmark.Iterations)
-                using (iteration.StartMeasurement())
-                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
-                    {
-                        ref string temp = ref span.DangerousGetPinnableReference();
-                    }
+            InvokeTestSpanDangerousGetPinnableReference<string>(length);
+        }
+
+        static void InvokeTestSpanDangerousGetPinnableReference<T>(int length)
+        {
+            var array = new T[length];
+
+            Invoke((int innerIterationCount) => TestSpanDangerousGetPinnableReference<T>(array, innerIterationCount),
+                "TestSpanDangerousGetPinnableReference<{0}>({1})", typeof(T).Name, length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void TestSpanDangerousGetPinnableReference<T>(T[] array, int iterationCount)
+        {
+            var sink = Sink<T>.Instance;
+            var span = new Span<T>(array);
+
+            for (int i = 0; i < iterationCount; i++)
+            {
+                ref T temp = ref span.DangerousGetPinnableReference();
+                sink.Data = temp;
+            }
         }
         #endregion
 
-        #region TestSpanIndex<T>
+        #region TestSpanIndexHoistable<T>
         [Benchmark(InnerIterationCount = BaseIterations)]
         [InlineData(1)]
         [InlineData(10)]
         [InlineData(100)]
         [InlineData(1000)]
-        public static void TestSpanIndexByte(int length)
+        public static void TestSpanIndexHoistableByte(int length)
         {
-            var array = new byte[length];
-            var span = new Span<byte>(array);
-            byte temp;
-            foreach (var iteration in Benchmark.Iterations)
-                using (iteration.StartMeasurement())
-                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
-                            temp = span[length/2];
+            InvokeTestSpanIndexHoistable<byte>(length);
+        }
+
+        [Benchmark(InnerIterationCount = BaseIterations)]
+        [InlineData(1)]
+        [InlineData(10)]
+        [InlineData(100)]
+        [InlineData(1000)]
+        public static void TestSpanIndexHoistableString(int length)
+        {
+            InvokeTestSpanIndexHoistable<string>(length);
+        }
+
+        static void InvokeTestSpanIndexHoistable<T>(int length)
+        {
+            var array = new T[length];
+
+            Invoke((int innerIterationCount) => TestSpanIndexHoistable<T>(array, length, innerIterationCount),
+                "TestSpanIndexHoistable<{0}>({1})", typeof(T).Name, length);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        static void TestSpanIndexHoistable<T>(T[] array, int length, int iterationCount)
+        {
+            var sink = Sink<T>.Instance;
+            var span = new Span<T>(array);
```

Comment:
Right, when I wrapped the tests in lambdas to make them callable from the command line, I lost the ability to put setup code in the same method as the kernel but outside the measurement loop.  So then there was a decision to be made about whether the spans created like this should be created inside the kernel method or created outside the loop and passed in.  With the SpanAPI tests, I opted to create the spans as locals in the kernel methods; passing the spans in would make measuring on win-x64 skewed since it would see them as byrefs (pending fixing #10049), so I thought this would best mirror the codegen we want to be tracking.  My expectation is (as you surmised in a later comment) that the loop's run-time will dwarf out the span creation time because it iterates so many times.

All that said, for IndexHoistable in particular, we'd hope to get to a point that the kernel loop doesn't even include the store... at that point I'd hope the running-time is effectively zero and the difference between creating the span here and not doing it in the array case would be too small to measure.

For any of these tests that you feel it's important, we could split them back out and have one method that gets invoked from Xunit (which could have a span local that gets initialized outside the timing loop) and another that gets invoked from the command line (which would look like this and we'd hope the codegen was similar enough).  My inclination is not to do that.

Regarding the choice of passing the spans into the method or not, while I made the spans local to the kernel methods for the SpanAPI tests, I made them parameters in the other version.  I just thought it would be good to have both approaches measured, and now hopefully the other tests will see a bump when #10049 gets implemented.